File: apps/frontend/src/types/moveQuality.ts
import type { Side } from './game';

/**
 * Move quality classification based on centipawn loss
 */
export type MoveQuality = 'best' | 'excellent' | 'good' | 'inaccuracy' | 'mistake' | 'blunder';

/**
 * Evaluation data for a single move
 */
export interface MoveEvaluation {
  moveIndex: number;
  san: string;
  evalBefore: number | null;
  evalAfter: number | null;
  cpLoss: number | null;
  quality: MoveQuality | null;
  isPlayerMove: boolean;
  side: Side;
  moveTimeMs?: number;
}

/**
 * Thresholds for move quality classification (in centipawns)
 * - Best: cpLoss = 0
 * - Excellent: cpLoss 1-20
 * - Good: cpLoss 21-50
 * - Inaccuracy: cpLoss 51-100
 * - Mistake: cpLoss 101-200
 * - Blunder: cpLoss > 200
 */
export const QUALITY_THRESHOLDS = {
  excellent: 20,
  good: 50,
  inaccuracy: 100,
  mistake: 200,
} as const;

/**
 * Border colors for each quality level
 */
export const QUALITY_COLORS: Record<MoveQuality, string> = {
  best: '#22c55e',
  excellent: '#4ade80',
  good: '#a3e635',
  inaccuracy: '#facc15',
  mistake: '#f97316',
  blunder: '#ef4444',
};

/**
 * Classify move quality based on centipawn loss
 * Returns null if cpLoss is null (evaluation not yet computed)
 */
export function classifyMoveQuality(cpLoss: number | null): MoveQuality | null {
  if (cpLoss === null) return null;

  if (cpLoss <= 0) return 'best';
  if (cpLoss <= QUALITY_THRESHOLDS.excellent) return 'excellent';
  if (cpLoss <= QUALITY_THRESHOLDS.good) return 'good';
  if (cpLoss <= QUALITY_THRESHOLDS.inaccuracy) return 'inaccuracy';
  if (cpLoss <= QUALITY_THRESHOLDS.mistake) return 'mistake';
  return 'blunder';
}

File: apps/mobile/lib/models/move_quality.dart
import 'package:flutter/painting.dart';

import 'game_types.dart';

enum MoveQuality { best, excellent, good, inaccuracy, mistake, blunder }

class QualityThresholds {
  static const int excellent = 20;
  static const int good = 50;
  static const int inaccuracy = 100;
  static const int mistake = 200;
}

const Map<MoveQuality, Color> qualityColors = {
  MoveQuality.best: Color(0xFF22C55E),
  MoveQuality.excellent: Color(0xFF4ADE80),
  MoveQuality.good: Color(0xFFA3E635),
  MoveQuality.inaccuracy: Color(0xFFFACC15),
  MoveQuality.mistake: Color(0xFFF97316),
  MoveQuality.blunder: Color(0xFFEF4444),
};

MoveQuality? classifyMoveQuality(double? cpLoss) {
  if (cpLoss == null) return null;
  if (cpLoss <= 0) return MoveQuality.best;
  if (cpLoss <= QualityThresholds.excellent) return MoveQuality.excellent;
  if (cpLoss <= QualityThresholds.good) return MoveQuality.good;
  if (cpLoss <= QualityThresholds.inaccuracy) return MoveQuality.inaccuracy;
  if (cpLoss <= QualityThresholds.mistake) return MoveQuality.mistake;
  return MoveQuality.blunder;
}

class MoveEvaluation {
  final int moveIndex;
  final String san;
  final double? evalBefore;
  final double? evalAfter;
  final double? cpLoss;
  final MoveQuality? quality;
  final bool isPlayerMove;
  final Side side;

  const MoveEvaluation({
    required this.moveIndex,
    required this.san,
    this.evalBefore,
    this.evalAfter,
    this.cpLoss,
    this.quality,
    required this.isPlayerMove,
    required this.side,
  });
}
